File: frontend/src/lib/workHour.js
export function computeWorkHourCompletion(rows, groupBy) {
  const keyMap = {
    function: 'Function Name',
    company: 'Comapny Name',
    location: 'Job Location',
  }
  const groupKey = keyMap[groupBy]
  const members = new Map()
  const presentCount = new Map()
  const odCount = new Map()
  const shiftHoursSum = new Map()
  const workHoursSum = new Map()
  const completedCount = new Map()
  const totalCount = new Map()

  const monthOf = (dateStr) => {
    if (!dateStr) return ''
    const m1 = String(dateStr).match(/(20\d{2})[-/](\d{1,2})/)
    if (m1) return `${m1[1]}-${String(m1[2]).padStart(2, '0')}`
    const m2 = String(dateStr).match(/(\d{1,2})[-/](\d{1,2})[-/](20\d{2})/)
    if (m2) return `${m2[3]}-${String(m2[2]).padStart(2, '0')}`
    const monthNames = {
      jan: 1, feb: 2, mar: 3, apr: 4, may: 5, jun: 6,
      jul: 7, aug: 8, sep: 9, sept: 9, oct: 10, nov: 11, dec: 12,
      january: 1, february: 2, march: 3, april: 4, june: 6, july: 7,
      august: 8, september: 9, october: 10, november: 11, december: 12,
    }
    const lower = String(dateStr).toLowerCase()
    const yearMatch = lower.match(/(20\d{2})/)
    const monthMatch = lower.match(/(jan|feb|mar|apr|may|jun|jul|aug|sep|sept|oct|nov|dec|january|february|march|april|june|july|august|september|october|november|december)/)
    if (yearMatch && monthMatch) {
      const year = yearMatch[1]
      const month = String(monthNames[monthMatch[1]]).padStart(2, '0')
      return `${year}-${month}`
    }
    return String(dateStr)
  }

  const timeToHours = (timeStr) => {
    if (!timeStr) return 0.0
    const s = String(timeStr).trim()
    const parts = s.split(/[:.]/)
    if (parts.length >= 2) {
      try {
        const h = parseInt(parts[0], 10)
        const m = parseInt(parts[1], 10)
        const sec = parts.length > 2 ? parseInt(parts[2], 10) : 0
        return h + m / 60.0 + sec / 3600.0
      } catch (e) {
        return 0.0
      }
    }
    return 0.0
  }

  for (const r of rows) {
    const month = monthOf(r['Attendance Date'] || '')
    const groupVal = String(r[groupKey] || '')
    const emp = String(r['Employee Code'] || r['Name'] || '')
    const flag = String(r['Flag'] || '').trim()
    const shiftIn = String(r['Shift In Time'] || '').trim()
    const shiftOut = String(r['Shift Out Time'] || '').trim()
    const inTime = String(r['In Time'] || '').trim()
    const outTime = String(r['Out Time'] || '').trim()

    const k = `${month}||${groupVal}`
    if (emp) {
      if (!members.has(k)) members.set(k, new Set())
      members.get(k).add(emp)
    }
    if (flag === 'P') {
      presentCount.set(k, (presentCount.get(k) || 0) + 1)
    }
    if (flag === 'OD') {
      odCount.set(k, (odCount.get(k) || 0) + 1)
    }

    const shiftInH = timeToHours(shiftIn)
    const shiftOutH = timeToHours(shiftOut)
    const inH = timeToHours(inTime)
    const outH = timeToHours(outTime)

    const shiftHrs = shiftOutH > shiftInH ? Math.max(0, shiftOutH - shiftInH) : 0
    const workHrs = outH > inH ? Math.max(0, outH - inH) : 0

    if (shiftHrs > 0 || workHrs > 0) {
      shiftHoursSum.set(k, (shiftHoursSum.get(k) || 0) + shiftHrs)
      workHoursSum.set(k, (workHoursSum.get(k) || 0) + workHrs)
      totalCount.set(k, (totalCount.get(k) || 0) + 1)
      if ((flag === 'P' || flag === 'OD') && workHrs >= shiftHrs && shiftHrs > 0) {
        completedCount.set(k, (completedCount.get(k) || 0) + 1)
      }
    }
  }

  const results = []
  for (const [k, memberSet] of members.entries()) {
    const [month, group] = k.split('||')
    const present = presentCount.get(k) || 0
    const od = odCount.get(k) || 0
    const shiftTotal = shiftHoursSum.get(k) || 0
    const workTotal = workHoursSum.get(k) || 0
    const completed = completedCount.get(k) || 0
    const total = totalCount.get(k) || 0
    const completionPct = (present + od) > 0 ? Number(((completed / (present + od)) * 100).toFixed(2)) : 0
    results.push({
      month,
      group,
      members: memberSet.size,
      present,
      od,
      shift_hours: Number(shiftTotal.toFixed(2)),
      work_hours: Number(workTotal.toFixed(2)),
      completed,
      completion_pct: completionPct,
    })
  }
  results.sort((a, b) => (a.month + a.group).localeCompare(b.month + b.group))
  return results
}

File: backend/app/services/work_hour.py
from __future__ import annotations

from typing import Dict, Any, List
from sqlalchemy.orm import Session
from sqlalchemy import select
from collections import defaultdict
import re

from ..models import UploadedRow


def _get_company_short_name(company_name: str) -> str:
    """Convert company name to short code."""
    company_map = {
        "Confidence Batteries Limited": "CBL",
        "Confidence Infrastructure PLC.": "CIPLC",
        "Confidence Steel Export Limited": "CSEL",
    }
    return company_map.get(company_name, company_name)


def _extract_month(date_str: str) -> str:
    if not date_str:
        return ""
    s = str(date_str)
    m = re.search(r"(20\d{2})[-/](\d{1,2})", s)
    if m:
        return f"{m.group(1)}-{int(m.group(2)):02d}"
    m = re.search(r"(\d{1,2})[-/](\d{1,2})[-/](20\d{2})", s)
    if m:
        return f"{m.group(3)}-{int(m.group(2)):02d}"
    month_map = {
        'jan':1,'feb':2,'mar':3,'apr':4,'may':5,'jun':6,
        'jul':7,'aug':8,'sep':9,'sept':9,'oct':10,'nov':11,'dec':12,
        'january':1,'february':2,'march':3,'april':4,'june':6,'july':7,'august':8,'september':9,'october':10,'november':11,'december':12,
    }
    lower = s.lower()
    ym = re.search(r"(20\d{2})", lower)
    mm = re.search(r"(jan|feb|mar|apr|may|jun|jul|aug|sep|sept|oct|nov|dec|january|february|march|april|june|july|august|september|october|november|december)", lower)
    if ym and mm:
        return f"{ym.group(1)}-{month_map[mm.group(1)]:02d}"
    return s


def _time_to_hours(time_str: str) -> float:
    if not time_str:
        return 0.0
    s = str(time_str).strip()
    parts = re.split(r'[:.]', s)
    if len(parts) >= 2:
        try:
            h = int(parts[0])
            m = int(parts[1])
            sec = int(parts[2]) if len(parts) > 2 else 0
            return h + m / 60.0 + sec / 3600.0
        except:
            pass
    return 0.0


def _compute_duration_hours(start_str: str, end_str: str) -> float:
    """Compute duration in hours, handling overnight shifts."""
    start_h = _time_to_hours(start_str)
    end_h = _time_to_hours(end_str)
    if start_h == 0.0 or end_h == 0.0:
        return 0.0
    # Handle overnight shifts (e.g., 22:00 to 06:00)
    if end_h < start_h:
        end_h += 24.0
    return max(0, end_h - start_h)


def compute_work_hour_completion(db: Session, group_by: str) -> List[Dict[str, Any]]:
    key_map = {
        "function": "Function Name",
        "company": "Comapny Name",
        "location": "Job Location",
    }
    group_key = key_map.get(group_by)
    if not group_key:
        raise ValueError("Invalid group_by")

    rows = db.execute(select(UploadedRow.data)).scalars().all()

    members = defaultdict(set)
    present_count = defaultdict(int)
    od_count = defaultdict(int)
    shift_hours_sum = defaultdict(float)
    work_hours_sum = defaultdict(float)
    completed_count = defaultdict(int)
    total_count = defaultdict(int)

    for r in rows:
        if not isinstance(r, dict):
            continue
        month = _extract_month(str(r.get("Attendance Date", "")))
        
        # For function-wise, combine Company - Function
        if group_by == "function":
            company_name = str(r.get("Comapny Name", "")).strip()
            function_name = str(r.get("Function Name", "")).strip()
            company_short = _get_company_short_name(company_name)
            if company_short and function_name:
                group_val = f"{company_short} - {function_name}"
            elif function_name:
                group_val = function_name
            else:
                group_val = company_short or "Unknown"
        else:
            group_val = str(r.get(group_key, ""))
        
        emp_code = str(r.get("Employee Code", "")).strip()
        emp_name = str(r.get("Name", "")).strip()
        member_id = emp_code or emp_name

        flag = str(r.get("Flag", "")).strip()
        shift_in = str(r.get("Shift In Time", "")).strip()
        shift_out = str(r.get("Shift Out Time", "")).strip()
        in_time = str(r.get("In Time", "")).strip()
        out_time = str(r.get("Out Time", "")).strip()

        key = (month, group_val)
        if member_id:
            members[key].add(member_id)
        if flag == "P":
            present_count[key] += 1
        if flag == "OD":
            od_count[key] += 1

        shift_hrs = _compute_duration_hours(shift_in, shift_out)
        work_hrs = _compute_duration_hours(in_time, out_time)

        # Only sum if we have valid shift hours
        if shift_hrs > 0:
            shift_hours_sum[key] += shift_hrs
            work_hours_sum[key] += work_hrs
            total_count[key] += 1
            if (flag == "P" or flag == "OD") and work_hrs >= shift_hrs and shift_hrs > 0:
                completed_count[key] += 1

    results = []
    for (month, group_val), member_set in members.items():
        key = (month, group_val)
        present = present_count.get(key, 0)
        od = od_count.get(key, 0)
        shift_total = shift_hours_sum.get(key, 0)
        work_total = work_hours_sum.get(key, 0)
        completed = completed_count.get(key, 0)
        total = total_count.get(key, 0)
        completion_pct = round((completed / (present + od) * 100.0), 2) if (present + od) > 0 else 0.0
        results.append({
            "month": month,
            "group": group_val,
            "members": len(member_set),
            "present": present,
            "od": od,
            "shift_hours": round(shift_total, 2),
            "work_hours": round(work_total, 2),
            "completed": completed,
            "completion_pct": completion_pct,
        })

    results.sort(key=lambda x: (x["month"], x["group"]))
    return results

